### DISCUSSION TITLE
Configurable maximum transmission frequency of specified message

### DISCUSSION THREAD

**Discussion by ffgamelife** [foxglove/community#1046] [General]
For high-frequency messages, visualization may only require a small part of them. High-frequency message communication will occupy a large amount of bandwidth; is it possible to specify the maximum transmission frequency of messages and filter the messages on the server side, like what [rosboard](https://github.com/dheera/rosboard/blob/main/rosboard/rosboard.py#L354) does?

**Comment by defunctzombie** [comment]
Is this referring to configuring this for Foxglove websocket? Or some other data source?

**Reply by ffgamelife** [reply]
Yes, expect to downsample the frequency of messages transmitted over websocket, regardless of the frequency of the original ros messages.
It is best to configure ros-foxglove-bridge through stdio client request, such as adding the max_freq field to the subscribe item of the protocol.

**Reply by phiabp** [reply]
Yes this would be a useful feature to allow high bandwidth streams such as IMU to be down sampled before being transmitted - reducing the demand on the network connection.

I.e. IMU is sampled at 1kHz on the Robot (which may be required for control), but Foxglove WebSocket only sends every 1/100th message.

Does this already exist or is this being considered for future improvements?

**Reply by amacneil** [reply]
When in live mode, would it be more helpful if we automatically decided the sampling rate?

E.g. a very small send buffer, if messages are not being delivered to the viz client fast enough we drop enough messages to keep the latency low?

This seems like helpful behavior for panels like Raw Message which are only going to display the latest message anyway. Maybe for plots having a more specific sampling rate would help? Or maybe the sampling rate is a "max" but we will still go lower than that rate if the send buffer cannot keep up?

**Reply by phiabp** [reply]
I think having deterministic down-sampling configurable on a topic-by-topic basis would be the best approach.

We’ve seen that using a single small global send buffer causes issues: if it’s set too small, larger topics (e.g., an H.264 video feed) get dropped while smaller, high-frequency topics still make it through. Unless you’re suggesting per-topic buffer settings?

Another reason for deterministic control is that certain messages or topics—such as event-driven messages that may only be sent once—shouldn't be dropped.
